IPIP-499: UnixFS CID Profiles

mishmosh opened the pull request:
Currently, CIDs can be generated with a variety of settings and optimizations for chunking, DAG width, and more. This means the same file can yield multiple, different CIDs depending on which tools and settings are used, and it is not possible to reliably reproduce or verify the CID. 

This proposal introduces profiles for IPFS CIDs. Profiles explicitly define CID version, hash algorithm, chunk size, DAG width, layout, and other parameters. They can be used to verify data across implementations, provide recommended settings depending on retrieval performance goals, and more.

mishmosh committed:
Create ipip-0000.md

mishmosh committed:
Update and rename ipip-0000.md to ipip-0499.md

2color commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,102 @@
+
+Each profile must specify the following characteristics:
+
+1. CID version (CIDv0 or CIDv1)
+2. Hash algorithm
+3. Chunk size
+4. DAG width
+5. DAG layout
```
I'd add the following parameters:
- HAMT threshold (when a block will be converted into a HAMT)
- HAMT fanout, aka HAMT dag width (all known examples use 256)
- Raw blocks (vs dag-pb file blocks)
- Are empty directories added by default? 
-

bumblefudge replied:
```suggestion
1. CID version (currently only CIDv0 or CIDv1)
2. Hash algorithm
3. UnixFS Chunk size (explicitly set, not contextual/reactive to input)
4. UnixFS directory DAG width
5. UnixFS directory DAG layout
6. HAMT directory DAG threshold
7. HAMT directory DAG width
8. Leaf Envelope (historically dag-pb, raw now)
9. Allow empty directories
```

bumblefudge commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,102 @@
+3. Chunk size
+4. DAG width
+5. DAG layout
+6. Required
```
I think there's a word missing here 😅

bumblefudge commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,102 @@
+| Chunk size  | 1MiB          | 256KiB                      | 1MiB             | 1MiB                  | not specified |
+| DAG width   | 1024          | 174 (but it's complicated*) | 1024             | 174                   | not specified |
+| DAG layout  | balanced      | balanced                    | balanced         | balanced              | not specified |
+
```
This is a good informational thing to show defaults are different, but what probably matters most to both end-users and implementers/backlog-managers is whether defaults can be overridden per input (e.g. by CLI flag) or per environment (e.g. by configuration/env var), or both.  It's totally fine for conformance to either profile to require extra work and not be the default, but I'm worried that even the affordance to override defaults for many of these settings lingers indefinitely in backlogs, or get more onerous to implement as time goes by!

lidel replied:
Good point, HAMT size threshold should be added.

fysa https://github.com/ipfs/kubo/pull/10774 aims to expose  `Import.UnixFSHAMTDirectorySizeThreshold` and I think we should agree how `cid-v1-2025` should look like: if we want to keep HAMT chunks at 256KiB, or if we want to align them with file chunk size which is 1MiB.

lidel commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,102 @@
+
+Additional profiles can be added at a future date. Profile names may be chosen from the names of any botanical tree with compound leaves.
+
+|             | Helia default | Kubo default                | Storacha default | "test-cid-v1" profile | DASL          |
```
This is only one of Kubo presets – basically "CIDv0" as CIDv1 + raw leaves. 
https://github.com/ipfs/kubo/pull/10774 renames `test-cid-v1` to `legacy-cid-v1` to make it more clear (I'll merge this comment once it lands)

```suggestion
|             | Helia default | Kubo default                | Storacha default | Kubo `legacy-cid-v1` profile | DASL          |
```

lidel committed:
add extra attributes proposed in review

Co-authored-by: Bumblefudge <bumblefudge@learningproof.xyz>

lidel committed:
incorporate kubo#10774

Import.* config params for controlling DAG width were added in:
https://github.com/ipfs/kubo/pull/10774

lidel commented:
Thank you for kicking this off, and filling initial state.

I've incorporated specific "dag width" settings for `File`, `Directory` and `HAMTDirectory` nodes, 
and updated the table to reflect state from https://github.com/ipfs/kubo/pull/10774 
and profiles that exist in Kubo `master` branch: `legacy-cid-v0`,  `test-cid-v1` and `test-cid-v1-wide`:
- https://github.com/ipfs/kubo/blob/master/config/profile.go#L268-L307

Next:
- [ ] agree what "`cid-2025`" profile should look like
  - this will be new default in "Kubo v1.0"
  - we have `test-cid-v1` and `test-cid-v1-wide` in Kubo as potential candidates
- [ ] switch to PR from local branch (so we have build preview)
- [ ] figure out how to render the information (currently the table is not supported by https://github.com/ipfs/spec-generator)

bumblefudge committed:
Merge branch 'main' into patch-1

2color commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,107 @@
+
+#### Empty directories
+
+Decision if empty directories should be included is left out of scope.
```
@lidel I think we should include this, because the goal of this spec is to make UnixFS encoding deterministic when using the profile, and this stands in the way of this, potentially rendering this whole effort futile if not included. 

I think the profile should dictate whether empty directories are included. This should also be configurable such that you can adjust to your own needs.

jaller94 commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,107 @@
+
+### User benefit
+
+Reliable, deterministic CIDs allow independent verification of content across tools and ipmlementations.
```
```suggestion
Reliable, deterministic CIDs allow independent verification of content across tools and implementations.
```

SethDocherty commented:
@jaller94 / @mishmosh, I've noticed a discrepancy in CID generation using the new `test-cid-v1-wide` profile and the CIDs generated with [Singularity](https://github.com/data-preservation-programs/singularity). I made a note of this in the following [post](https://discuss.ipfs.tech/t/should-we-profile-cids/18507/40?u=sethdocherty) on discuss. ipfs.tech along with this [post](https://discuss.ipfs.tech/t/fixed-defaultlinksperblock-value-and-the-ipfs-add-command/17635) outlining how Singularity has set `dag-width`.

This could be an edge case, and I wanted to document it here as I'd expect the same CID to be generated. Would either of you be able to point me to a repo within the `ipfs` umbrella to post this as an issue?

2color committed:
Update src/ipips/ipip-0499.md

Co-authored-by: Christian Paul <info@jaller.de>

2color committed:
add daniel as editor

2color committed:
edit summary and motivation

2color committed:
edit summary

2color committed:
edit parameters and design

2color committed:
edit user benefit and compatibility

2color committed:
refine parameters and introduce a named profile

2color commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,131 @@
+
+The profiles define a set of parameters that affect the resulting CID. These parameters are based on the UnixFS specification and are used to generate the CID for a given file tree. The parameters include:
+
+1. CID version (currently only CIDv0 or CIDv1)
```
Good point. I think we should dictate the multibase too. The final string representation of the CID matters here.

rvagg replied:
dagpb encodes the whole cid bytes, so version matters

2color commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,131 @@
+The profiles define a set of parameters that affect the resulting CID. These parameters are based on the UnixFS specification and are used to generate the CID for a given file tree. The parameters include:
+
+1. CID version (currently only CIDv0 or CIDv1)
+1. Hash function
```
That's right. I think this IPIP should dictate the hash function for all nodes in the DAG to meet the goal of CID determinism. That's the point of the profile. This is a subset of UnixFS after all.

2color commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,131 @@
+1. CID version (currently only CIDv0 or CIDv1)
+1. Hash function
+1. UnixFS chunk size
+1. UnixFS DAG layout (e.g. balanced, trickle)
```
> We probably need to fix the unixfs spec as it doesn't really say how the layout works. 

Yeah, maybe a point to add to https://github.com/ipfs/specs/pull/331

> For a start, they DAG layout is not part of the unixfs "format", but rather unixfs protobufs are embedded in merkledag-pb protobufs, that could be cbors or jsons. So either we set that in stone (unixfs dag == protobuf-pb) or we need to be explicit in profiles.

I think I understand. Do you mean that the DAG layout is a function of the `PBNode.Links` which are on the dag-pb layer, i.e. outer envelope of UnixFS?

> Then there is the part about the subtleties of the layout. i.e. the trickle parameters, the balancing of the balanced dag etc, which are also badly specified. It may be that now you can get away with UnixFS DAG width for configuring existing DAG layouts, but I would imagine layouts with more options than the width.

Right. We should refine this section a bit more and keep it open ended. But how do we avoid a combinatorial explosion to account for all possibilities? 

This IPIP introduces a single named profile, `unixfs-2025` and the minimum necessary parameters to define it. 

My question to you is if you actually think it's necessary to already introduce profiles for other DAG layouts?

rvagg commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,131 @@
+1. Leaf Envelope: either `dag-pb` or `raw`
+1. Whether empty directories are included in the DAG
+  - Some implementations apply filtering before merkleizing filesystem entries in the DAG.
+
```
couple of other things to consider?

* Directory wrapping at the top level (for just files, kubo has an option to wrap in a directory so you get file metadata)
* Presence and accurate setting of `Tsize` - at one point we were going to deprecate this field for some cases, although I think all our encoders now do it properly, you could just mandate this in the spec though -- all valid profiles must properly encode `Tsize`.

2color replied:
I've added this as a parameter. 

According to the latest version of https://github.com/ipfs/specs/pull/331/files, the calculation is done as follows:

> To compute the Tsize of a child DAG, sum the length of the dag-pb outside message binary length and the blocksizes of all nodes in the child DAG.

If calculated according to this, does it make accurate?

rvagg replied:
sounds about right, I remember there being some nuance in exactly what's included in the size calculation, making it not super stable if you get it slightly wrong (as we did for some variants in go-unixfsnode for a while)

2color commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,131 @@
+1. `HAMTDirectory` fanout (must be a power of 2)
+1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on an estimate of the block size by counting the size of PNNode.Links
+1. Leaf Envelope: either `dag-pb` or `raw`
+1. Whether empty directories are included in the DAG
+  - Some implementations apply filtering before merkleizing filesystem entries in the DAG.
```
This is motivated by Git's default behaviour which ignores empty directories.

darobin replied:
Do we have existing implementations that support filtering differently on all of these? Because unless we do, I would really rather not specify all possible variants. And I agree with @bumblefudge: let's have two behaviours if possible, and punt to the UnixFS spec for how to describe them.

2color committed:
Apply suggestions from code review

Co-authored-by: Hector Sanjuan <code@hector.link>

2color committed:
edit based on hector's feedback

2color committed:
Apply suggestions from code review

Co-authored-by: Rod Vagg <rod@vagg.org>

2color committed:
add multibase encoding

2color committed:
address feedback from rvagg

bumblefudge commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,131 @@
+1. UnixFS DAG layout (e.g. balanced, trickle)
+1. UnixFS DAG width (max number of links per `File` node)
+1. `HAMTDirectory` fanout (must be a power of 2)
+1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on an estimate of the block size by counting the size of PNNode.Links
```
wait is kubo dynamically assigning HAMT Directory threshold, currently? i was assuming this was a static number!

achingbrain replied:
If you want to be exact you have to take into account any non-zero value fields in the serialized root UnixFS metadata since these affect the block size.

It's quite possible that Kubo will produce a HAMT block that's too big with a certain combination of directory entry names if someone has also changed the encoded directory's default mtime or whatever, probably because the "should-I-shard" feature pre-dates Kubo's ability to add UnixFSv1.5 metadata to things.

Really there's no need to estimate anything - it's trivial to count the actual bytes that a block will take up and then shard if necessary.

2color commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,131 @@
+1. UnixFS chunk size
+1. UnixFS DAG layout (e.g. balanced, trickle)
+1. UnixFS DAG width (max number of links per `File` node)
+1. `HAMTDirectory` fanout (must be a power of 2)
```
Bit-width is a bit more intuitive imo, because it ties more deeply to how the HAMT is implemented.

darobin commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,132 @@
+1. UnixFS DAG layout (e.g. balanced, trickle etc...)
+1. UnixFS DAG width (max number of links per `File` node)
+1. `HAMTDirectory` bitwidth, i.e. the number of bits determines the fanout of the `HAMTDirectory` (default bitwidth is 8 == 256 leaves).
+1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on an estimate of the block size by counting the size of PNNode.Links
```
```suggestion
1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on an estimate of the block size by counting the size of PNNode.Links. We do not include details about the estimation algorithm as we do not encourage implementations to support it.
```

achingbrain replied:
I don't think we should be estimating the block size as it's trivial to calculate it exactly.  Can we not just define this (and punt to the spec for the details) to make it less hand-wavey?

```suggestion
1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on the final size of the serialized form of the [PBNode protobuf message](https://specs.ipfs.tech/unixfs/#dag-pb-node) that represents the directory.
```

rvagg commented:
Hey, I'd love to be able to reference this, even if it's in "draft" form, could we just merge it and continue to iterate on top of it to get it right?

mishmosh committed:
Update ipip-0499.md

Fixed outdated references, consistent profile names, streamlined Summary and Motivation sections.

rvagg commented on src/ipips/ipip-0499.md:
```diff
@@ -0,0 +1,132 @@
+| `HAMTDirectory` fanout        | 256 blocks                                              |
+| `HAMTDirectory` threshold     | 256KiB (estimated by counting the size of PBNode.links) |
+| Leaves                        | raw                                                     |
+| Empty directories             | TODO                                                    |
```
```suggestion
| Empty directories             | TODO                                                    |
| Hidden entities               | TODO                                                    |
```

mishmosh committed:
Update src/ipips/ipip-0499.md

Co-authored-by: Rod Vagg <rod@vagg.org>

mishmosh committed:
Update src/ipips/ipip-0499.md

Co-authored-by: Rod Vagg <rod@vagg.org>

lidel committed:
Update src/ipips/ipip-0499.md

Co-authored-by: Rod Vagg <rod@vagg.org>

mishmosh committed:
fix typo (the the)

lidel committed:
Merge branch 'main' into patch-1

lidel committed:
feat(ipip-0499): add chunking algorithm and align profile tables

- add chunking algorithm parameter to both tables (fixed-size)
- add hidden entities row to legacy profiles table
- ensures both unixfs-2025 and legacy tables cover same parameters

lidel committed:
fix(ipip-0499): correct kubo legacy profile

- rename kubo-legacy-2015 to kubo-legacy-2025
- clarify (v0.39 default) instead of (kubo default)
- fix leaves value: dag-pb (UnixFSRawLeaves=False in legacy-cid-v0)

lidel committed:
fix(ipip-0499): document legacy profile filtering behavior

clarify empty directories and hidden entities handling with precise
terminology based on kubo v0.39, helia, and storacha implementations:

- `included`: always in DAG, no option to exclude (kubo/helia empty dirs)
- `excluded`: never in DAG, no option to include (storacha empty dirs)
- `opt-in`: excluded by default, flag to include (all hidden entities)
- `opt-out`: included by default, flag to exclude

add terminology note to explain these terms

lidel committed:
fix(ipip-0499): note that legacy table includes non-UnixFS implementations

lidel committed:
feat(ipip-0499): add implementation versions to legacy profiles table

add "Based on" row with package/tool versions and kubo profile names

lidel committed:
fix(ipip-0499): update HAMTDirectory threshold and clean up parameters

- unixfs-2025: mark threshold as TODO, prefer Helia's block size approach
- unixfs-2025: note kubo needs opt-out flag for empty directories
- legacy profiles: add estimation method to kubo profiles
- parameters section: add backticks, clarify threshold estimation methods

lidel committed:
feat(ipip-0499): document symlink handling in profiles

- add Symlinks parameter to UnixFS parameters list
- add Symlinks row to unixfs-2025 (TODO) and legacy profiles tables
- kubo: preserved, helia/storacha: followed, dasl: not specified
- add terminology for preserved/followed with UnixFS spec reference
- clarify kubo --dereference-args behavior

lidel reviewed (changes requested):
Quick update: I've pushed several commits addressing feedback and gaps in the document:

## Resolved / Research done

- Documented Syhmlink behavior as suggested by @icidasset 
  - Only Kubo 0.39 preserves symlinks, everything else dereferences on the fly by default, turning symlinks into real files and directories a symlink pointed at
- Added Chunking algorithm row to both profile tables for completeness
- Fixed kubo-legacy-2025 profile: corrected Leaves from raw to dag-pb (verified against kubo v0.39 legacy-cid-v0 profile where UnixFSRawLeaves=false)
- Documented filtering behavior with clear terminology:
  - included: always in DAG (no option to exclude)
  - excluded: always excluded (no option to include)
  - opt-in: excluded by default, flag to include (e.g., --hidden)
  - opt-out: included by default, flag to exclude
- Added Based on row with implementation versions and kubo profile names (legacy-cid-v0, test-cid-v1, test-cid-v1-wide)
- Clarified HAMTDirectory threshold estimation methods in the parameters section: link count (naive), PBNode.Links size (name + CID), or full dag-pb block size (most accurate)
- Noted that legacy table includes non-UnixFS implementations (DASL) in Summary section
- Added estimation method suffix (est:links[name+cid]) to kubo profiles in legacy table


## Remaining TODOs in `unixfs-2025`

| Parameter               | Status                                         |
|-------------------------|------------------------------------------------|
| HAMTDirectory threshold | TODO - fix kubo: likely based on full block size estimation (Helia approach) |
| Empty directories       | TODO - use kubo? needs opt-out flag  + `Import.*`               |
| Hidden entities         | TODO - use kubo? needs opt-in flag  + `Import.*` |
| Symlinks                | TODO - use kubo? needs flag + `Import.*` for controlling if all symlinks in imported directory tree are preserved or dereferenced) |
| Test fixtures           | TODO - reuse kubo: will reuse once kubo has them for `*-2025` profiles                                          |

Other:
- Look into https://github.com/data-preservation-programs/singularity/issues/525


## Implementation Plan (Kubo 0.40, ETA 2026 Q1)

To finalize this IPIP, Kubo needs to support additional `Import.*` configuration flags for:
1. Empty directories: opt-out flag to exclude them from DAG
2. Hidden files: already has `--hidden`, just need to wire it up from config
3. HAMTDirectory threshold: configurable to support both legacy estimation (name + CID size) and Helia-style full block size calculation

Test fixtures will likely be included in the same Kubo PR that adds these missing features.

I also think we may replace two `kubo-2025` and `kubo-2025-wide` profiles with a single one, that makes decision on what remains narrow and what is wide, but will update once Kubo changes land. (now that we have convention of doing IPIPs with profiles, we can always course-correct in `-202

lidel committed:
fix(ipip-0499): clarify HAMTDirectory threshold calculation methods

recommend full serialized PBNode size, link to dag-pb spec

ref: https://github.com/ipfs/specs/pull/499#discussion_r2523340313

lidel committed:
fix(ipip-0499): update metadata and add contributors

- rename to UnixFS CID Profiles
- add lidel as editor
- add thanks section with PR reviewers

lidel committed:
feat(ipip-0499): document HAMTDirectory threshold estimation methods

- add `links-count`, `links-bytes`, `block-bytes` estimation methods
- fill unixfs-2025 profile: 256KiB (block-bytes), empty dirs included (opt-out), hidden opt-in, symlinks preserved
- update legacy profiles table with cleaner method names
- clarify profile naming allows YYYY or YYYY-MM suffix
- add reference from unixfs.md to IPIP-499 for threshold methods
- fix broken markdown links and formatting
- clarify compliance requirements: MUST support unixfs-2025, MAY support legacy profiles

lidel committed:
chore: bump spec-generator to 1.7.0

adds GFM table support for markdown

lidel committed:
feat(ipip-0499): restructure document and rename profiles

- rename `unixfs-2025` to `unixfs-v1-2025` for clarity on CID version
- add `unixfs-v0-2015` legacy profile for backward compatibility with kubo
- move divergence analysis from separate section into Motivation
- clarify motivation: CIDs are verifiable, problem is DAG construction variance
- consolidate three problems into two: broken hash semantics, verification overhead
- add Mode and Mtime parameters to all profile tables
- improve UnixFS parameters section with inline links and explanations
- add balanced vs trickle DAG layout descriptions
- update unixfs.md to reference IPIP-499 in a note block with full URL

lidel committed:
feat(ipip-0499): document efficiency benefits of modern profile parameters

add user benefit explaining why 1 MiB chunks with 1024 links per node
results in shallower DAG trees, fewer total nodes, faster seeking,
and reduced DHT announcement overhead compared to legacy 256 KiB/174 params

lidel committed:
Merge branch 'main' into patch-1

lidel committed:
feat(ipip-0499): add singularity to divergence table

include singularity as example showing balanced layout has implementation
variants that affect CID determinism for large files:
- document balanced-packed DAG layout variant
  (https://github.com/data-preservation-programs/singularity/issues/525)
- note boxo defaults for HAMT parameters
- note rclone defaults for hidden files and symlinks

lidel committed:
feat(ipip-0499): add test fixtures section with deterministic CIDs

- add verification process for testing profile compliance
- add test vectors for unixfs-v0-2015 profile (5 CIDs)
- add test vectors for unixfs-v1-2025 profile (5 CIDs)
- document HAMT threshold behavior (> not >=)
- document balanced vs trickle DAG layout with ASCII diagrams
- add trickle layout CID for reader compatibility testing
- reference existing test vectors from UnixFS spec (empty dir, symlink)

lidel committed:
fix(ipip-0499): update test fixtures with chunk threshold vectors

add chunk size threshold test vectors for precise boundary testing:
- file-at-chunk: file exactly at chunk size (single block)
- file-over-chunk: file +1 byte over chunk size (2 blocks)

update file-over-max-links CIDs to use +1 byte threshold instead of
+1 chunk, enabling more precise DAG rebalancing boundary tests.

lidel committed:
refactor(ipip-0499): restructure Motivation section

- move HAMT sharding history to Motivation, add HAMT switch comparison to profiles
- rename Divergence section to Divergences across ecosystem
- merge Observed differences into balanced-packed, promote subsections to h3
- move Why sections under Compatibility

lidel commented:
@parkan Not for me to say, it is up to the project. The difference (`balanced` vs `balanced-packed`) is documented in the IPIP, so Singularity could adopt it or state "we made concious decision to use `balanced-packed` as defined in IPIP-499" and call it a day.

Whether matching future Kubo 1.0 CIDs (the `unixfs-v1-2025` profile) matters depends on your use case. This isn't a holy grail: many projects will keep using specialized DAG layouts for their specific needs (e.g., [WebRecorder](https://webrecorder.net) uses a [custom chunker for WARC files](https://specs.webrecorder.net/wacz-ipfs/latest/#warc-file-chunking)).

mishmosh committed:
Merge branch 'main' into patch-1

achingbrain commented on src/ipips/ipip-0499.md:
```diff
@@ -115,16 +115,47 @@ Name introduced by this IPIP for [Singularity][singularity]'s variant. Groups pr
+
+- 2026-01: [boxo#1088](https://github.com/ipfs/boxo/pull/1088) fixed threshold comparison from `>=` to `>`, aligning with JS implementation and documentation. Shipped in Kubo 0.40.
+
+Timeline of JavaScript implementation changes:
```
js-ipfs did link counting, the PR to js-ipfs-unixfs did block-bytes but then [switched](https://github.com/ipfs/js-ipfs-unixfs/pull/171#issuecomment-1424040277) to [link-bytes](https://github.com/ipfs/js-ipfs-unixfs/pull/171/changes/288910ffe735ba5eee9a5007491b9b7ffa4567ab) to be consistent with kubo.

lidel committed:
chore(ipip-0499): mark as ratified

- ipip-0499: update status from proposal to ratified
- ipip-0499: update date to 2026-03-05
- unixfs.md: update date to 2026-03-05

Luflosi commented:
Thank you for the very quick response and for fixing my problem so thoroughly.

Thank you also for explaining why simply using zeroes would not work, I did not consider that.
